src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/6.sql

```diff
@@ -0,0 +1,1944 @@
+--
+CREATE PROCEDURE dbo.AcquireExportJobs
+    @jobHeartbeatTimeoutThresholdInSeconds bigint,
+    @maximumNumberOfConcurrentJobsAllowed int
+AS
+    SET NOCOUNT ON
+    SET XACT_ABORT ON
+
+    SET TRANSACTION ISOLATION LEVEL SERIALIZABLE
+    BEGIN TRANSACTION
+
+    -- We will consider a job to be stale if its timestamp is smaller than or equal to this.
+    DECLARE @expirationDateTime dateTime2(7)
+    SELECT @expirationDateTime = DATEADD(second, -@jobHeartbeatTimeoutThresholdInSeconds, SYSUTCDATETIME())
+
+    -- Get the number of jobs that are running and not stale.
+    -- Acquire and hold an exclusive table lock for the entire transaction to prevent jobs from being created, updated or deleted during acquisitions.
+    DECLARE @numberOfRunningJobs int
+    SELECT @numberOfRunningJobs = COUNT(*) FROM dbo.ExportJob WITH (TABLOCKX) WHERE Status = 'Running' AND HeartbeatDateTime > @expirationDateTime
+
+    -- Determine how many available jobs we can pick up.
+    DECLARE @limit int = @maximumNumberOfConcurrentJobsAllowed - @numberOfRunningJobs;
+
+    DECLARE @availableJobs TABLE (Id varchar(64) COLLATE Latin1_General_100_CS_AS NOT NULL, JobVersion binary(8) NOT NULL)
+
+    -- Get the available jobs, which are export jobs that are queued or stale.
+    -- Older jobs will be prioritized over newer ones.
+    INSERT INTO @availableJobs
+    SELECT TOP(@limit) Id, JobVersion
+    FROM dbo.ExportJob
+    WHERE (Status = 'Queued' OR (Status = 'Running' AND HeartbeatDateTime <= @expirationDateTime))
+    ORDER BY HeartbeatDateTime
+
+    DECLARE @heartbeatDateTime datetime2(7) = SYSUTCDATETIME()
+
+    -- Update each available job's status to running both in the export table's status column and in the raw export job record JSON.
+    UPDATE dbo.ExportJob
+    SET Status = 'Running', HeartbeatDateTime = @heartbeatDateTime, RawJobRecord = JSON_MODIFY(RawJobRecord,'$.status', 'Running')
+    OUTPUT inserted.RawJobRecord, inserted.JobVersion
+    FROM dbo.ExportJob job INNER JOIN @availableJobs availableJob ON job.Id = availableJob.Id AND job.JobVersion = availableJob.JobVersion
+
+    COMMIT TRANSACTION
+GO
+
+/*************************************************************
+    Search Parameter Status Information
+**************************************************************/
+
+CREATE TYPE dbo.SearchParamTableType_1 AS TABLE
+(
+    Uri varchar(128) COLLATE Latin1_General_100_CS_AS NOT NULL,
+    Status varchar(10) NOT NULL,
+    IsPartiallySupported bit NOT NULL
+)
+
+GO
+
+/*************************************************************
+    Stored procedures for search parameter information
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     GetSearchParamStatuses
+--
+-- DESCRIPTION
+--     Gets all the search parameters and their statuses.
+--
+-- RETURN VALUE
+--     The search parameters and their statuses.
+--
+CREATE PROCEDURE dbo.GetSearchParamStatuses
```

Comment:
>GetSearchParamStatuses [](start = 21, length = 22)

I'm looking into other store procedures, and they at least have some other logic other than just plain select.
I'm curious, why it's a store procedure rather than just plain sql code?

Reply:
Good point. I think this could go either way since the performance is the same (had to double check this and ended up reading through a post about it [here](https://docs.microsoft.com/en-us/previous-versions/dotnet/articles/ms973918(v=msdn.10))). I think it's nice to have this as a stored procedure because it makes the `SqlServerSearchParameterStatusDataStore` file a bit cleaner/more consistent. What do you think?

Reply:
I think amount lines of code in `SqlServerSearchParameterStatusDataStore` would be the same, it's just instead of populating command via generated code you directly specify command as string. But that would also reduce amount of code everywhere else.

If I was a DBA, I would probably advocate what everything need to be store procedure and triggers, but as developer I don't like idea of splitting business logic and pushing it into database layer. It can be often justified by performance reasons, but that's not the case.

Technically I would ask some of the elders in the project, what is common strategy for store procedures vs sql code. I don't have much background knowledge regarding data access layer to be right person to make a call, but I can express opinions :)

Reply:
Nice, you make a good point. @brendankowitz thoughts?

Reply:
One thing to note: the `GetExportJobById` is another stored proc that only contains a single `SELECT` statement

Reply:
Fair enough, precedent law is on your side 😄